src/libraries/System.Text.Json/System.Text.Json.SourceGeneration/JsonSourceGeneratorHelper.cs

```diff
@@ -0,0 +1,559 @@
+                Diagnostics.Add(Diagnostic.Create(_generatedTypeClass, Location.None, new string[] { root.Name, className }));
+
+                // Add generated typeinfo for current traversal.
+                Types.Add(type, source.ToString());
+            }
+            else
+            {
+                Diagnostics.Add(Diagnostic.Create(_failedToGenerateTypeClass, Location.None, new string[] { root.Name, className }));
+
+                // If not successful remove it from found types hashset and add to failed types list.
+                seenTypes.Remove(type);
+                _failedTypes.Add(type);
+            }
+
+            return isSuccessful;
+        }
+
+        // Call recursive type generation if unseen type and check for success and cycles.
+        void GenerateForMembers(Type root, Type currentType, HashSet<Type> seenTypes, ref bool isSuccessful)
+        {
+            // If new type, recurse.
+            if (IsNewType(currentType, seenTypes))
+            {
+                bool wasSuccessful = GenerateClassInfo(root, seenTypes, currentType.Name, currentType);
+                isSuccessful &= wasSuccessful;
+
+                if (!wasSuccessful)
+                {
+                    Diagnostics.Add(Diagnostic.Create(_failedToAddNewTypesFromMembers, Location.None, new string[] { root.Name, currentType.Name }));
+                }
+            }
+        }
+
+        // Check if current type is supported to be iterated over.
+        private bool IsSupportedType(Type type)
+        {
+            if (type is TypeWrapper typeWrapper)
+            {
+                if (typeWrapper.IsIEnumerable)
+                {
+                    // todo: Add more support to collections.
+                    if (!typeWrapper.IsIList)
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        // Returns name of types traversed that can be looked up in the dictionary.
+        public List<Type> GenerateClassInfo(string className, Type rootType)
+        {
+            HashSet<Type> foundTypes = new HashSet<Type>();
+            GenerateClassInfo(rootType, foundTypes, className, rootType);
+            return foundTypes.ToList();
+        }
+
+        private Type[] GetTypesToGenerate(Type type)
+        {
+            if (type.IsArray)
+            {
+                return new Type[] { type.GetElementType() };
+            }
+            if (type.IsGenericType)
+            {
+                return type.GetGenericArguments();
+            }
+
+            return new Type[] { type };
+        }
+
+        private bool IsNewType(Type type, HashSet<Type> foundTypes) => (
+            !Types.ContainsKey(type) &&
+            !foundTypes.Contains(type) &&
+            !s_simpleTypes.Contains(type));
+
+        private bool AddImportsToTypeClass(Type type, PropertyInfo[] properties, FieldInfo[] fields, StringBuilder source)
```

Comment:
Seems suspect that we are passing `PropertyInfo[]`, `StringBuilder` etc. in nearly every method. Can we consolidate them in a struct & have a concept of a "current" type being processed (which would be an instance member on `JsonSourceGenerator`)? See `GenerationClassFrame` in an earlier prototype: https://github.com/layomia/jsonconvertergenerator/blob/master/JsonConverterGenerator.cs#L12-L31